### PR TITLE
Fused GDN attention support

### PR DESCRIPTION
**Pull request by ksivaman**
- [x] I, the PR author, have personally reviewed every line of this PR.

# What does this PR do?

 
TransformerEngine integrated the fused GDN kernels in PR https://github.com/NVIDIA/TransformerEngine/pull/3351. This PR adds plumbing for those changes in mcore.

## Issue tracking

For PRs from open-source community contributors:

- **New features**: a linked issue is **required**. Please open a [feature request](https://github.com/NVIDIA/Megatron-LM/issues/new?template=feature_request.md) and reference it here before submitting the PR.
- **Small updates (bug fixes, minor improvements)**: a linked issue is **recommended** and will accelerate the PR review process.

Linked issue: 

### Pre-checks

- [x] I have added relevant unit tests
- [x] I have added relevant functional tests
- [ ] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [ ] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

Feel free to message or comment @NVIDIA/mcore-oncall to help accelerate your merge into main. The less complex your PR is, the faster it will be approved and merged!

All PRs start as **draft**. If you open a non-draft PR, it will be automatically converted to draft.

#### Step 1: Mark PR as "Ready for Review"

1. When your PR is ready, click **Ready for Review**.
2. An oncall reviewer is auto-assigned and expert reviewers are notified based on your changes.
   - Some PRs may jump straight to step 2. This is determined by `.github/CODEOWNERS`.

:warning: Only mark as ready once merge-conflicts are resolved and the CI is passing.
Final Review might get declined if these requirements are not fulfilled.

#### Step 2: Final Review

For PRs that change `megatron/core`, once all expert reviewers have approved, the `Final Review` label is applied **automatically** and final reviewers are assigned.

For PRs outside `megatron/core`, this step is skipped.

#### Step 3: Approved

Once all required reviewers have approved, the `Approved` label is applied **automatically**.

### Merge

Any member of [mcore-engineers](https://github.com/orgs/NVIDIA/teams/mcore-engineers) will be able to merge your PR.

### CODE REVIEW

File: megatron/core/ssm/gated_delta_net/gdn.py [hunk 1]
```diff
@@ -58,7 +57,71 @@ def _setup_variant_attrs(self):
         if self.config.deterministic_mode:
             self.gated_delta_rule = torch_chunk_gated_delta_rule
         else:
-            self.gated_delta_rule = chunk_gated_delta_rule
```
Comment (cspades) [inline megatron/core/ssm/gated_delta_net/gdn.py:72]: Still think we should have the TE knob control this instead of replacing the FLA variant.

Either way, some parity / convergence tests with CP and THD are in order as well! I would like to know if there are any CuDNN discrepancies and if so how mild they are (e.g. 1e-3, 1e-4, etc.) for future ref.

File: megatron/core/ssm/gated_delta_net/gdn.py [hunk 2]
```diff
@@ -58,7 +57,71 @@ def _setup_variant_attrs(self):
...
+            try:
+                from transformer_engine.pytorch import DotProductAttention
+            except ImportError as exc:
+                raise ImportError(
+                    "GDN requires TransformerEngine with fused Gated DeltaNet support."
+                ) from exc
```
Comment (cspades) [inline megatron/core/ssm/gated_delta_net/gdn.py]: This needs to be more specific - perhaps checking for `transformer_engine.pytorch.attention.dot_product_attention.gdn` or something specifically pointing at GDN support in the DPA API (like checking for `g` or `beta` or some other signature config for GDN) would be better. And it needs a TE guard (`HAVE_TE`) as well!

File: megatron/core/ssm/gated_delta_net/gdn.py [hunk 3]
```diff
@@ -58,7 +57,71 @@ def _setup_variant_attrs(self):
...
+            attention_mask=None,
+            qkv_format=qkv_format,
+            cu_seqlens_q=cu_seqlens,
+            cu_seqlens_kv=cu_seqlens,
+            g=g.float(),
+            beta=beta.float(),
+            initial_state=initial_state,
+            output_final_state=output_final_state,
+            use_qk_l2norm_in_kernel=use_qk_l2norm_in_kernel,
+        )
```
Comment (cspades) [inline megatron/core/ssm/gated_delta_net/gdn.py]: You could implement this trivial interface [`GatedDeltaRuleInterface`](https://github.com/NVIDIA/Megatron-LM/blob/main/megatron/core/ssm/gated_delta_net/common.py#L70) for consistency so swapping between implementations is seamless.

**Review by cspades — APPROVED** [review]

File: megatron/core/ssm/gated_delta_net/common.py [hunk 4]
```diff
@@ -528,6 +528,9 @@ def sharded_state_dict(self, prefix="", sharded_offsets=(), metadata=None, tp_gr
                     tp_group=tp_group,
                     dp_cp_group=metadata['dp_cp_group'],
                 )
+            elif name == "core_attention":
+                # Skip core_attention
```
Comment (jaredcasper) [inline megatron/core/ssm/gated_delta_net/common.py]: This feels scary... why is this here? Maybe add a bit more context to the comment as this is fairly unusual.

File: megatron/core/transformer/transformer_config.py [hunk 5]
```diff
@@ -367,6 +367,12 @@ class TransformerConfig(ModelParallelConfig):
     - An integer N: Represents a (N-1):N ratio, meaning (N-1) LA layers for every 1 SDPA layer
     - A list that defines a custom pattern, e.g.: [1,1,1,0,1,1,1,0,1,1,1,0]"""
 
+    gdn_kernel_backend: Literal["auto", "transformer_engine", "fla"] = "auto"
```
Comment (jaredcasper) [inline megatron/core/transformer/transformer_config.py]: No "auto"! In mcore there is no such thing as a "requested backend". You put in what you want and you get it or it errors out.

File: megatron/core/ssm/gated_delta_net/gdn.py [hunk 6]
```diff
@@ -65,12 +96,87 @@ def _setup_variant_attrs(self):
...
 
         if self.config.deterministic_mode:
+            self.gdn_backend = "torch"
             self.gated_delta_rule = torch_chunk_gated_delta_rule
-        else:
+            return
+
+        requested_backend = self.config.gdn_kernel_backend
+        if requested_backend == "fla" or (requested_backend == "auto" and not HAVE_TE_GDN):
+            self.gdn_backend = "fla"
```
Comment (jaredcasper) [inline megatron/core/ssm/gated_delta_net/gdn.py:84]: See comment in TransformerConfig. We do not select which code to run based on runtime availability.

File: megatron/core/ssm/gated_delta_net/gdn.py [hunk 7]
```diff
@@ -65,12 +96,87 @@ def _setup_variant_attrs(self):
 
         self.dt_bias_dim = self.num_v_heads_local_tp
         self.a_log_dim = self.num_v_heads_local_tp
+        self.chunk_size = 64
 
         if self.config.deterministic_mode:
+            self.gdn_backend = "torch"
```
Comment (jaredcasper) [inline megatron/core/ssm/gated_delta_net/gdn.py:79]: Don't silently change the backend! If they ask for deterministic mode and a non-deterministic backend that's an error.

File: megatron/core/ssm/gated_delta_net/gdn.py [hunk 8]
```diff
@@ -65,12 +96,87 @@ def _setup_variant_attrs(self):
...
+        v: torch.Tensor,
+        g: torch.Tensor,
+        beta: torch.Tensor,
+        initial_state: torch.Tensor | None = None,
+        output_final_state: bool = False,
+        use_qk_l2norm_in_kernel: bool = False,
+        cu_seqlens: torch.Tensor | None = None,
+        **kwargs,
+    ) -> tuple[torch.Tensor, torch.Tensor | None]:
+        """Adapt Megatron GDN tensors to TransformerEngine's fused GDN API."""
```
Comment (jaredcasper) [inline megatron/core/ssm/gated_delta_net/gdn.py]: This feels like it should be part of a wrapper in the transformer_engine extension, not a private method here.

**Review by zhongbozhu — APPROVED** [review]

**Review by Phlip79 — APPROVED** [review]

**Review by Phlip79 — CHANGES_REQUESTED** [review]
Need https://github.com/NVIDIA/TransformerEngine/pull/3351 merged

File: tests/unit_tests/ssm/test_gated_delta_net.py [hunk 9]
```diff
@@ -325,3 +726,282 @@ def test_cp1_still_validates_total(self, mock_gdn):
         actual = torch.tensor([0, 500, 1000], dtype=torch.int32)
         with pytest.raises(ValueError, match="does not match"):
             mock_gdn._resolve_cu_seqlens(None, actual, 1008, "cu_seqlens_q", cp_size=1)
+
+
+@pytest.mark.parametrize("sequence_packing", [False, True])
```
Comment (zhongbozhu) [inline tests/unit_tests/ssm/test_gated_delta_net.py:731]: please make sure this unit test launches on GB200 with `pytest.mark.launch_on_gb200`

File: tests/unit_tests/ssm/test_gated_delta_net.py [hunk 10]
```diff
@@ -281,6 +520,168 @@ def test_jit_compiled_helpers(self):
             assert (g <= 0).all()
             assert (beta >= 0).all() and (beta <= 1).all()
 
+    def test_gpu_forward_thd_padding_correctness(self):
```
Comment (zhongbozhu) [inline tests/unit_tests/ssm/test_gated_delta_net.py:523]: same comment for adding `pytest.mark.launch_on_gb200`

File: tests/unit_tests/ssm/test_gated_delta_net.py [hunk 11]
```diff
@@ -281,6 +520,168 @@ def test_jit_compiled_helpers(self):
...
+        with pytest.raises(ValueError, match="does not match"):
+            self.gdn(hidden_states_thd, None, packed_seq_params=padded_mismatch_params)
+
+        # D) actual mismatch branch without *_padded: should raise.
+        actual_mismatch_params = make_test_packed_seq_params(cu_seqlens=[0, 32, 64, 96, 129])
+        with pytest.raises(ValueError, match="does not match"):
+            self.gdn(hidden_states_thd, None, packed_seq_params=actual_mismatch_params)
+
+
+@pytest.mark.skipif(not HAVE_TE_GDN, reason="TransformerEngine GDN is not available.")
```
Comment (zhongbozhu) [inline tests/unit_tests/ssm/test_gated_delta_net.py:586]: same^

**Review by zhongbozhu — CHANGES_REQUESTED** [review]

File: megatron/core/ssm/gated_delta_net/gdn.py (hunk 8 above)
Comment (ksivaman) [reply megatron/core/ssm/gated_delta_net/gdn.py]: Moved to the extensions wrapper.

File: megatron/core/ssm/gated_delta_net/gdn.py (hunk 7 above)
Comment (ksivaman) [reply megatron/core/ssm/gated_delta_net/gdn.py:79]: Addressed! Now we raise an error for unsupported configurations -- for e.g. deterministic TE backend with GDN attention

File: megatron/core/ssm/gated_delta_net/gdn.py (hunk 6 above)
Comment (ksivaman) [reply megatron/core/ssm/gated_delta_net/gdn.py:84]: Done

File: megatron/core/transformer/transformer_config.py (hunk 5 above)
Comment (ksivaman) [reply megatron/core/transformer/transformer_config.py]: Done!

File: megatron/core/ssm/gated_delta_net/common.py (hunk 4 above)
Comment (ksivaman) [reply megatron/core/ssm/gated_delta_net/common.py]: Added explainer comment

File: tests/unit_tests/ssm/test_gated_delta_net.py [hunk 12]
```diff
@@ -325,3 +728,283 @@ def test_cp1_still_validates_total(self, mock_gdn):
...
+        linear_num_value_heads=8,
+        num_layers=1,
+        normalization="RMSNorm",
+        use_cpu_initialization=True,
+        layernorm_zero_centered_gamma=True,
+        num_attention_heads=8,
+        activation_func=F.silu,
+        bf16=True,
+        experimental_attention_variant="gated_delta_net",
+        gdn_kernel_backend="fla",
```
Comment (zhongbozhu) [inline tests/unit_tests/ssm/test_gated_delta_net.py:759]: If the backend is FLA, how can we test the newly added TE kernel?

**Review by zhongbozhu — CHANGES_REQUESTED** [review]

**Review by Phlip79 — APPROVED** [review]

**Comment by ericharper** [comment]
/ok to test: [3fa2dc1](https://github.com/NVIDIA/Megatron-LM/pull/6645/commits/3fa2dc139c4293d4cff94dd3c4af236d8b3231b8)

**Review by zhongbozhu — APPROVED** [review]
Thanks for the fix!

File: tests/unit_tests/ssm/test_gated_delta_net.py (hunk 12 above)
Comment (zhongbozhu) [reply tests/unit_tests/ssm/test_gated_delta_net.py]: This issue has been resolved and it exposed a cudnn-fe bug when head dim is 32 multiple: https://github.com/NVIDIA/cudnn-frontend/pull/994

**Comment by ericharper** [comment]
/ok to test [23826a3](https://github.com/NVIDIA/Megatron-LM/pull/6645/commits/23826a3403b3aee95494210fa595030f30b414ce)

**Review by jaredcasper — APPROVED** [review]
